### DISCUSSION TITLE
How to create mixed JS-Rust projects with NAPI-RS?

### DISCUSSION THREAD

**Discussion by chriskilding** [napi-rs/napi-rs#1245] [Q&A]
At the moment, creating an all-Rust module with NAPI-RS is the main use case. But in some cases we need to have some of our module's code in Javascript, and some in Rust.

The reasons for this might include:

- Keeping certain operations within JS to avoid the performance hit of copying to/from Rust
- The need to call Node API functions

Does NAPI-RS support mixed JS-Rust projects today, and if so how?

**Comment by TheBrenny** [comment]
So I think I'm running into a similar problem... Instead of exposing all my rust functions, I want to write a wrapper over the top, but for some parts, invoked the bindings directly. The way I'm planning on solving this is writing a wrapper that `require`s the bindings, and then re-exports them as required.

More or less, that's how you'll end up solving your problem: `require` the bindings, and re-export them, plus a few more - with your `main` being that "newer" script.

You can specify where you want to save your bindings by specifying the `--js "path/to/script.js` build argument:

```
> napi build --js "js_src/bindings.js --platform"
> cd js_src
> ls
bindings.js
bindings.d.js
index.js
index.d.js
```

Where `index.js` is _your_ script.
